pubtools/exodus/_tasks/push.py

```diff
@@ -0,0 +1,69 @@
+import logging
+import subprocess
+
+from pubtools.exodus.task import ExodusTask
+
+LOG = logging.getLogger("pubtools-exodus")
+LOG_FORMAT = "%(asctime)s [%(levelname)-8s] %(message)s"
+
+
+class ExodusPushTask(ExodusTask):
+    """Push a directory to the Exodus CDN"""
+
+    def __init__(self):
+        super(ExodusPushTask, self).__init__()
+        self.src = None
+        self.dest = None
+
+    def add_args(self):
+        super(ExodusPushTask, self).add_args()
+
+        self.parser.add_argument(
+            "--source",
+            help="The path to the directory to be pushed to the exodus CDN",
+        )
+        self.parser.add_argument(
+            "--dest",
+            help="The destination path to publish to on the exodus CDN",
+        )
+
+    def run(self):
+        LOG.debug("Exodus push begins")
+
+        if not self.publish:
+            self.publish = self.new_publish()
+            LOG.info("created exodus-gw publish: %s", self.publish["id"])
+
+        self.src = self.args.source
+        self.dest = (
+            self.args.dest
+            if ":" in self.args.dest
+            else "exodus:%s" % self.args.dest
+        )
+
+        cmd = [
+            "exodus-rsync",
+            "--exodus-publish",
+            self.publish["id"],
+            self.src,
+            self.dest,
+        ]
+        LOG.info(cmd)
+
+        with subprocess.Popen(
+            cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT
+        ) as proc:
+            out = proc.communicate()[0]
```

Comment:
This is not ideal because you won't get any output at all until the command completes.  So, for example, if it gets stuck for hours then the Pub task log would only show the exodus-rsync command, and then no further output at all.

It would be good to improve this.  Some options would be:

1. Don't use communicate, instead loop over the command's output.  Something like "for line in proc.stdout: ...". 

Note although there's a warning in the docs to "Use communicate() rather than .stdin.write, .stdout.read or .stderr.read", I think the warning is getting at cases where you try to deal with more than one of stdin/stdout/stderr as for example you might try to read from stdout, but the process might be trying to read from stdin first, so you deadlock.    I don't think it's an issue here as we only need to deal with proc.stdout.

2. Don't pass stdout/stderr to popen, then the output will stream in real time to the parent's stdout/stderr.

Unfortunately I'm not 100% sure this is going to work when it runs in Pub, we'd have to check it during testing.